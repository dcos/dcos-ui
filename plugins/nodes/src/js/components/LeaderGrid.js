import React from "react";
import { FormattedMessage } from "react-intl";
import moment from "moment";

import ConfigurationMap from "#SRC/js/components/ConfigurationMap";
import ConfigurationMapHeading from "#SRC/js/components/ConfigurationMapHeading";
import ConfigurationMapLabel from "#SRC/js/components/ConfigurationMapLabel";
import ConfigurationMapRow from "#SRC/js/components/ConfigurationMapRow";
import ConfigurationMapSection from "#SRC/js/components/ConfigurationMapSection";
import ConfigurationMapValue from "#SRC/js/components/ConfigurationMapValue";
import Loader from "#SRC/js/components/Loader";

function timeFromNow(time) {
  if (!time) {
    return null;
  }

  return moment(time * 1000).fromNow();
}

const Loading = () => <Loader size="small" type="ballBeat" />;

const ConfigurationRow = ({ keyValue, title, value }) => {
  const loadedValue = value || <Loading />;

  return (
    <ConfigurationMapRow key={keyValue}>
      <ConfigurationMapLabel>{title}</ConfigurationMapLabel>
      <ConfigurationMapValue>{loadedValue}</ConfigurationMapValue>
    </ConfigurationMapRow>
  );
};

export default function LeaderGrid({ leader }) {
  return (
    <div className="container">
      <ConfigurationMap>
        <ConfigurationMapHeading className="flush-top">
          Leader
        </ConfigurationMapHeading>
        <ConfigurationMapSection>
          <ConfigurationRow
            keyValue="leader"
            title="IP and Port"
            value={leader.hostPort}
          />

          <ConfigurationRow
            keyValue="region"
            title="Region"
            value={leader.region}
          />

          <ConfigurationRow
            keyValue="version"
            title={<FormattedMessage id="COMMON.VERSION" />}
            value={leader.version}
          />

          <ConfigurationRow
            keyValue="started"
            title={<FormattedMessage id="COMMON.STARTED" />}
            value={timeFromNow(leader.startTime)}
          />

          <ConfigurationRow
            keyValue="elected"
            title={<FormattedMessage id="COMMON.ELECTED" />}
            value={timeFromNow(leader.electedTime)}
          />
        </ConfigurationMapSection>
      </ConfigurationMap>
    </div>
  );
}
