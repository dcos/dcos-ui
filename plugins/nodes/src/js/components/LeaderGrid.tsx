import { Trans, DateFormat } from "@lingui/macro";
import * as React from "react";
import { ToggleContent } from "@dcos/ui-kit";

import ConfigurationMap from "#SRC/js/components/ConfigurationMap";
import ConfigurationMapHeading from "#SRC/js/components/ConfigurationMapHeading";
import ConfigurationMapLabel from "#SRC/js/components/ConfigurationMapLabel";
import ConfigurationMapRow from "#SRC/js/components/ConfigurationMapRow";
import ConfigurationMapSection from "#SRC/js/components/ConfigurationMapSection";
import ConfigurationMapValue from "#SRC/js/components/ConfigurationMapValue";
import Loader from "#SRC/js/components/Loader";
import DateUtil from "#SRC/js/utils/DateUtil";

const Loading = () => <Loader size="small" type="ballBeat" />;

const ConfigurationRow = ({ keyValue, title, value }) => {
  const loadedValue = value || <Loading />;

  return (
    <ConfigurationMapRow key={keyValue}>
      <ConfigurationMapLabel>{title}</ConfigurationMapLabel>
      <ConfigurationMapValue>{loadedValue}</ConfigurationMapValue>
    </ConfigurationMapRow>
  );
};

export default function LeaderGrid({ leader }) {
  return (
    <div className="container">
      <ConfigurationMap>
        <ConfigurationMapHeading className="flush-top">
          <Trans id="Leader" />
        </ConfigurationMapHeading>
        <ConfigurationMapSection>
          <ConfigurationRow
            keyValue="leader"
            title={<Trans id="IP and Port" />}
            value={leader.hostPort}
          />

          <ConfigurationRow
            keyValue="region"
            title={<Trans id="Region" />}
            value={leader.region}
          />

          <ConfigurationRow
            keyValue="version"
            title={<Trans id="Mesos Version" />}
            value={leader.version}
          />

          {/* L10NTODO: Relative time */}
          <ConfigurationRow
            keyValue="started"
            title={<Trans id="Started" />}
            value={
              <ToggleContent
                contentOn={DateUtil.msToRelativeTime(leader.startTime * 1000)}
                contentOff={
                  <DateFormat
                    value={new Date(leader.startTime * 1000)}
                    format={DateUtil.getFormatOptions("longMonthDateTime")}
                  />
                }
              />
            }
          />

          {/* L10NTODO: Relative time */}
          <ConfigurationRow
            keyValue="elected"
            title={<Trans id="Elected" />}
            value={
              <ToggleContent
                contentOn={DateUtil.msToRelativeTime(leader.electedTime * 1000)}
                contentOff={
                  <DateFormat
                    value={new Date(leader.electedTime * 1000)}
                    format={DateUtil.getFormatOptions("longMonthDateTime")}
                  />
                }
              />
            }
          />
        </ConfigurationMapSection>
      </ConfigurationMap>
    </div>
  );
}
