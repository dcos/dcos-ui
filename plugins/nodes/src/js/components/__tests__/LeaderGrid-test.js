import React from "react";
import renderer from "react-test-renderer";
import { IntlProvider } from "react-intl";
import enUS from "#SRC/js/translations/en-US.json";

import LeaderGrid from "../LeaderGrid";

const master = {
  hostPort: "127.1.2.3:8080",
  version: "2.9.1",
  electedTime: 1532340694.04573,
  startTime: 1232340694.04573,
  region: "us-east-1"
};

const TestableLeaderGrid = ({ leader }) => {
  return (
    <IntlProvider locale="en" messages={enUS}>
      <LeaderGrid leader={leader} />
    </IntlProvider>
  );
};

describe("LeaderGrid", function() {
  beforeEach(function() {
    Date.now = jest.fn(() => 1542340694);
  });

  afterEach(function() {
    Date.now.mockRestore();
  });

  it("renders with running status", function() {
    expect(
      renderer.create(<TestableLeaderGrid leader={master} />).toJSON()
    ).toMatchSnapshot();
  });
});
