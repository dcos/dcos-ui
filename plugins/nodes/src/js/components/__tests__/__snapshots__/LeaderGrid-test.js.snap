// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaderGrid renders with running status 1`] = `
<div
  className="container"
>
  <div
    className="configuration-map"
  >
    <h1
      className="detail-view-section-heading detail-view-section-heading-primary configuration-map-heading flush-top"
    >
      Leader
    </h1>
    <div
      className="configuration-map-section"
    >
      <div
        className="configuration-map-row table-row"
      >
        <div
          className="configuration-map-label"
        >
          IP and Port
        </div>
        <div
          className="configuration-map-value"
        >
          127.1.2.3:8080
        </div>
      </div>
      <div
        className="configuration-map-row table-row"
      >
        <div
          className="configuration-map-label"
        >
          Region
        </div>
        <div
          className="configuration-map-value"
        >
          us-east-1
        </div>
      </div>
      <div
        className="configuration-map-row table-row"
      >
        <div
          className="configuration-map-label"
        >
          <span>
            Version
          </span>
        </div>
        <div
          className="configuration-map-value"
        >
          2.9.1
        </div>
      </div>
      <div
        className="configuration-map-row table-row"
      >
        <div
          className="configuration-map-label"
        >
          <span>
            Started
          </span>
        </div>
        <div
          className="configuration-map-value"
        >
          in 39 years
        </div>
      </div>
      <div
        className="configuration-map-row table-row"
      >
        <div
          className="configuration-map-label"
        >
          <span>
            Elected
          </span>
        </div>
        <div
          className="configuration-map-value"
        >
          in 49 years
        </div>
      </div>
    </div>
  </div>
</div>
`;
